common4j/src/main/com/microsoft/identity/common/java/util/CommonURIBuilder.java

```diff
@@ -0,0 +1,162 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.util;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Map;
+
+import cz.msebera.android.httpclient.NameValuePair;
+import cz.msebera.android.httpclient.client.utils.URIBuilder;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import lombok.NonNull;
+
+/**
+ * Our URIBuilder.
+ * We want to make sure we never send duplicated parameters to the server.
+ * This is done by
+ * 1. disabling {@link cz.msebera.android.httpclient.client.utils.URIBuilder#addParameter(String, String)} and
+ *    {@link cz.msebera.android.httpclient.client.utils.URIBuilder#addParameters(List))}
+ * 2. adding {@link CommonURIBuilder#addParametersIfAbsent}
+ */
+public class CommonURIBuilder extends cz.msebera.android.httpclient.client.utils.URIBuilder {
+
+    public CommonURIBuilder(){
+        super();
+    }
+
+    public CommonURIBuilder(final String string) throws URISyntaxException {
+        super(string);
+    }
+
+    public CommonURIBuilder(final URI uri) {
+        super(uri);
+    }
+
+    @Override
+    public CommonURIBuilder addParameters(@NonNull final List<NameValuePair> nvps) {
+        throw new UnsupportedOperationException("This should never be used. Either use setParameter or addParametersIfAbsent");
+    }
+
+    @Override
+    public CommonURIBuilder addParameter(@NonNull final String param, @NonNull final String value) {
+        throw new UnsupportedOperationException("This should never be used. Either use setParameter or addParametersIfAbsent");
+    }
+
+    @Override
+    public CommonURIBuilder setParameters(List<NameValuePair> nvps) {
+        super.setParameters(nvps);
+        return this;
+    }
+
+    @Override
+    public CommonURIBuilder setParameters(NameValuePair... nvps) {
+        super.setParameters(nvps);
+        return this;
+    }
+
+    @Override
+    public CommonURIBuilder setParameter(String param, String value) {
+        super.setParameter(param, value);
+        return this;
+    }
```

Comment:
seems like we're just calling `super` - do we still need to override these?

Reply:
it's just to return CommonURIBuilder instead of URIBuilder

Reply:
but `CommandURIBuilder` is `URIBuilder` -  I guess we would need to cast in that case though. Fine.